File: demos/python/event_driven_teleop.py
import signal
import time

import numpy as np
import trossen_arm
import trossen_vr as vr

START_POSE = [0, np.pi / 3, np.pi / 6, np.pi / 5, 0, 0]
IDLE_POSE = [0, 0, 0, 0, 0, 0]

running = True


def signal_handler(sig, frame):
    global running
    running = False


signal.signal(signal.SIGINT, signal_handler)
signal.signal(signal.SIGTERM, signal_handler)

# Configuration
RIGHT_ARM_IP = "192.168.1.4"
LEFT_ARM_IP = "192.168.1.5"
SEND_RATE_HZ = 100.0
GRIPPER_MAX_M = 0.04
CMD_GOAL_TIME = 0.15


def relax_velocity_effort_limits(driver):
    """Widen the arm joints' velocity/effort fault *tolerance* to their max so normal
    teleop motion doesn't spuriously trip a fault. Genuine faults (e.g. a real position
    limit) still trip and are handled via ArmTeleopState.faulted rather than crashing.
    """
    limits = driver.get_joint_limits()
    for i in range(6):
        limits[i].velocity_tolerance = limits[i].velocity_max
        limits[i].effort_tolerance = limits[i].effort_max
    driver.set_joint_limits(limits)


# Robot setup
right_driver = trossen_arm.TrossenArmDriver()
right_driver.configure(
    trossen_arm.Model.wxai_v0,
    trossen_arm.StandardEndEffector.wxai_v0_leader,
    RIGHT_ARM_IP,
    False,
)
relax_velocity_effort_limits(right_driver)
right_driver.set_all_modes(trossen_arm.Mode.position)

left_driver = trossen_arm.TrossenArmDriver()
left_driver.configure(
    trossen_arm.Model.wxai_v0,
    trossen_arm.StandardEndEffector.wxai_v0_leader,
    LEFT_ARM_IP,
    False,
)
relax_velocity_effort_limits(left_driver)
left_driver.set_all_modes(trossen_arm.Mode.position)

print("Moving arms to start position")
right_driver.set_arm_positions(START_POSE, 2.0, True)
left_driver.set_arm_positions(START_POSE, 2.0, True)

# Network setup
net_config = vr.ReceiverConfig()
net_config.port = 9000
receiver = vr.NetworkManager(net_config)
receiver.start()


# Arm state management
class ArmTeleopState:
    def __init__(self):
        self.engaged = False
        self.T_offset = vr.Transform4D()
        self.last_vr_pose = vr.Pose6D()
        self.pose_valid = False
        # Set when a driver call raises (e.g. a joint limit). While True, this
        # arm is skipped entirely until recovered (button A - see handle_a()).
        self.faulted = False


right_state = ArmTeleopState()
left_state = ArmTeleopState()


def engage_arm(state, driver, label):
    if state.faulted or state.engaged or not state.pose_valid:
        return

    try:
        current_p = driver.get_cartesian_positions()
        robot_pose = vr.Pose6D()
        robot_pose.x, robot_pose.y, robot_pose.z = (
            current_p[0],
            current_p[1],
            current_p[2],
        )
        robot_pose.ax, robot_pose.ay, robot_pose.az = (
            current_p[3],
            current_p[4],
            current_p[5],
        )

        T_robot = vr.pose6d_to_transform4d(robot_pose)
        T_vr = vr.pose6d_to_transform4d(state.last_vr_pose)
        state.T_offset = T_robot * T_vr.inverse()
        state.engaged = True
    except Exception as e:
        print(f"{label}: FAULT, press A to recover: {e}")
        state.faulted = True


def disengage_arm(state):
    if not state.engaged:
        return
    state.engaged = False
    state.pose_valid = False


def send_arm_command(state, driver, cmd_goal_time, label):
    if state.faulted or not state.engaged or not state.pose_valid:
        return

    try:
        T_vr = vr.pose6d_to_transform4d(state.last_vr_pose)
        T_cmd = state.T_offset * T_vr
        cmd_pose = vr.transform4d_to_pose6d(T_cmd)
        goal = [
            cmd_pose.x,
            cmd_pose.y,
            cmd_pose.z,
            cmd_pose.ax,
            cmd_pose.ay,
            cmd_pose.az,
        ]
        driver.set_cartesian_positions(
            goal, trossen_arm.InterpolationSpace.cartesian, cmd_goal_time, False
        )
    except Exception as e:
        print(f"{label}: FAULT, press A to recover: {e}")
        state.faulted = True


def try_recover(state, driver, label, home_pose):
    """Attempt to clear a fault and prime the arm to resume smoothly. Returns True if
    the arm is (now) not faulted -- a no-op success if it wasn't faulted to begin with.
    """
    if not state.faulted:
        return True
    try:
        driver.clear_error()
        driver.set_all_modes(trossen_arm.Mode.position)
        print(f"{label}: moving to home before resuming teleop")
        driver.set_arm_positions(home_pose, 2.0, True)
        state.faulted = False
        # Force a fresh offset capture instead of resuming from the stale one,
        # so the arm ramps to the controller's current pose rather than jumping.
        state.engaged = False
        print(f"{label}: fault cleared, re-engage by moving the controller")
        return True
    except Exception as e:
        print(f"{label}: clear_error failed, still faulted: {e}")
        return False


# Track previous tracking state for automatic engage/disengage
prev_right_tracked = 0
prev_left_tracked = 0

# Event-driven teleop setup
teleop = vr.Teleop()


# Button B - exit
def handle_b():
    global running
    print("Exit requested via B button")
    running = False


# Button A (right controller) - recover either/both arms from a fault
def handle_a():
    if not right_state.faulted and not left_state.faulted:
        print("A pressed, but no arm is faulted")
        return
    try_recover(right_state, right_driver, "Right arm", START_POSE)
    try_recover(left_state, left_driver, "Left arm", START_POSE)


# Right trigger - right gripper
def handle_right_trigger(val):
    if right_state.faulted:
        return
    try:
        right_driver.set_gripper_position(val * GRIPPER_MAX_M, 0.0, False)
    except Exception as e:
        print(f"Right arm: FAULT, press A to recover: {e}")
        right_state.faulted = True


# Left trigger - left gripper
def handle_left_trigger(val):
    if left_state.faulted:
        return
    try:
        left_driver.set_gripper_position(val * GRIPPER_MAX_M, 0.0, False)
    except Exception as e:
        print(f"Left arm: FAULT, press A to recover: {e}")
        left_state.faulted = True


# Right controller pose - auto engage/disengage based on tracking
def handle_right_pose(pose):
    global prev_right_tracked
    right_state.last_vr_pose = pose
    right_state.pose_valid = True

    if not prev_right_tracked:
        # Auto-engage arm
        engage_arm(right_state, right_driver, "Right arm")
        print("Right arm ENGAGED")
    prev_right_tracked = 1


# Left controller pose - auto engage/disengage based on tracking
def handle_left_pose(pose):
    global prev_left_tracked
    left_state.last_vr_pose = pose
    left_state.pose_valid = True

    if not prev_left_tracked:
        # Auto-engage arm
        engage_arm(left_state, left_driver, "Left arm")
        print("Left arm ENGAGED")
    prev_left_tracked = 1


teleop.on_button_a(handle_a)
teleop.on_button_b(handle_b)
teleop.on_right_trigger(handle_right_trigger)
teleop.on_left_trigger(handle_left_trigger)
teleop.on_right_pose(handle_right_pose)
teleop.on_left_pose(handle_left_pose)

print(
    "Waiting for VR data... Hand/Grip trigger to engage, release to pause. "
    "Press A to recover from a fault, B to exit"
)

send_period = 1.0 / SEND_RATE_HZ
last_send = time.monotonic()
last_status = vr.ConnectionStatus.Disconnected

while running:
    # Monitor connection status
    current_status = receiver.get_connection_status()
    if current_status != last_status:
        if current_status == vr.ConnectionStatus.Connecting:
            print("Connecting...")
        elif current_status == vr.ConnectionStatus.Connected:
            print(f"Connection established ({receiver.get_message_frequency():.1f} Hz)")
        elif current_status == vr.ConnectionStatus.Degraded:
            print(
                f"Connection degraded (low frequency: {receiver.get_message_frequency():.1f} Hz)"
            )
        elif current_status == vr.ConnectionStatus.Disconnected:
            print("Connection lost (timeout)")
        last_status = current_status

    frame = receiver.latest_frame()
    if frame is None:
        time.sleep(0.001)
        continue

    # Dispatch to event handlers
    teleop.dispatch(frame)

    # Check for tracking release - disengage
    right_tracked = frame.right_controller.is_tracked != 0
    left_tracked = frame.left_controller.is_tracked != 0

    if not right_tracked and prev_right_tracked:
        disengage_arm(right_state)
        print("Right arm PAUSED")
        prev_right_tracked = 0
    if not left_tracked and prev_left_tracked:
        disengage_arm(left_state)
        print("Left arm PAUSED")
        prev_left_tracked = 0

    # Rate-limited command sending
    now = time.monotonic()
    if now - last_send >= send_period:
        send_arm_command(right_state, right_driver, CMD_GOAL_TIME, "Right arm")
        send_arm_command(left_state, left_driver, CMD_GOAL_TIME, "Left arm")
        last_send = now

receiver.stop()
print("\nShutting down...")
print("Moving arms to idle position")
try:
    right_driver.set_arm_positions(IDLE_POSE, 2.0, True)
except Exception as e:
    print(f"Right arm: could not move to idle (still faulted?): {e}")
try:
    left_driver.set_arm_positions(IDLE_POSE, 2.0, True)
except Exception as e:
    print(f"Left arm: could not move to idle (still faulted?): {e}")

File: demos/cpp/event_driven_teleop.cpp
#include <cmath>
#include <csignal>

#include <array>
#include <chrono>
#include <iostream>
#include <string>
#include <thread>
#include <unordered_map>
#include <vector>

#include "libtrossen_arm/trossen_arm.hpp"
#include "trossen_vr/trossen_vr.hpp"

static volatile std::sig_atomic_t running = 1;

void signal_handler(int) {
    running = 0;
}

struct ArmModelConfig {
    trossen_arm::Model model;
    trossen_arm::EndEffector end_effector;
};

// Supported arm models and their matching end effector
const std::unordered_map<std::string, ArmModelConfig> kModelConfigs = {
    {"wxai_v0", {trossen_arm::Model::wxai_v0, trossen_arm::StandardEndEffector::wxai_v0_follower}},
    {"pro", {trossen_arm::Model::pro, trossen_arm::StandardEndEffector::pro_base}},
};

bool parse_model_arg(const std::string& value, ArmModelConfig& config) {
    auto it = kModelConfigs.find(value);
    if (it == kModelConfigs.end()) {
        return false;
    }
    config = it->second;
    return true;
}

struct ArmTeleopState {
    bool engaged = false;
    trossen_vr::Transform4D T_offset;
    trossen_vr::Pose6D last_vr_pose;
    bool pose_valid = false;
};

void engage_arm(ArmTeleopState& state, trossen_arm::TrossenArmDriver& driver) {
    if (state.engaged || !state.pose_valid) return;

    auto current_p = driver.get_cartesian_positions();
    trossen_vr::Pose6D robot_pose;
    robot_pose.x = current_p[0]; robot_pose.y = current_p[1]; robot_pose.z = current_p[2];
    robot_pose.ax = current_p[3]; robot_pose.ay = current_p[4]; robot_pose.az = current_p[5];

    trossen_vr::Transform4D T_robot = trossen_vr::pose6d_to_transform4d(robot_pose);
    trossen_vr::Transform4D T_vr = trossen_vr::pose6d_to_transform4d(state.last_vr_pose);
    state.T_offset = T_robot * T_vr.inverse();
    state.engaged = true;
}

void disengage_arm(ArmTeleopState& state) {
    if (!state.engaged) return;
    state.engaged = false;
    state.pose_valid = false;
}

void send_arm_command(ArmTeleopState& state, trossen_arm::TrossenArmDriver& driver,
                      double cmd_goal_time) {
    if (!state.engaged || !state.pose_valid) return;

    trossen_vr::Transform4D T_vr = trossen_vr::pose6d_to_transform4d(state.last_vr_pose);
    trossen_vr::Transform4D T_cmd = state.T_offset * T_vr;
    trossen_vr::Pose6D cmd_pose = trossen_vr::transform4d_to_pose6d(T_cmd);
    std::array<double, 6> goal{cmd_pose.x, cmd_pose.y, cmd_pose.z,
                              cmd_pose.ax, cmd_pose.ay, cmd_pose.az};
    driver.set_cartesian_positions(
        goal, trossen_arm::InterpolationSpace::cartesian,
        cmd_goal_time, false
    );
}

int main(int argc, char** argv) {
    std::signal(SIGINT, signal_handler);
    std::signal(SIGTERM, signal_handler);

    // Parse --model-left / --model-right (wxai_v0 or pro)
    std::string model_left_arg = "wxai_v0";
    std::string model_right_arg = "wxai_v0";
    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];
        if (arg == "--model-left" && i + 1 < argc) {
            model_left_arg = argv[++i];
        } else if (arg == "--model-right" && i + 1 < argc) {
            model_right_arg = argv[++i];
        } else if (arg == "-h" || arg == "--help") {
            std::cout << "Usage: " << argv[0]
                      << " [--model-left wxai_v0|pro] [--model-right wxai_v0|pro]" << std::endl;
            return 0;
        }
    }

    ArmModelConfig left_arm_config, right_arm_config;
    if (!parse_model_arg(model_left_arg, left_arm_config)) {
        std::cerr << "Unknown model '" << model_left_arg << "' for --model-left. "
                  << "Valid options: wxai_v0, pro" << std::endl;
        return 1;
    }
    if (!parse_model_arg(model_right_arg, right_arm_config)) {
        std::cerr << "Unknown model '" << model_right_arg << "' for --model-right. "
                  << "Valid options: wxai_v0, pro" << std::endl;
        return 1;
    }

    // Configuration
    const std::string right_arm_ip = "192.168.1.4";
    const std::string left_arm_ip = "192.168.1.5";
    const double send_rate_hz = 100.0;
    const double gripper_max_m = 0.04;
    const double cmd_goal_time = 0.15;

    trossen_vr::ReceiverConfig net_config;
    net_config.port = 9000;

    const std::vector<double> START_POSE = {0, M_PI/3, M_PI/6, M_PI/5, 0, 0};
    const std::vector<double> IDLE_POSE  = {0, 0, 0, 0, 0, 0};

    // Robot setup
    trossen_arm::TrossenArmDriver right_driver;
    trossen_arm::TrossenArmDriver left_driver;

    right_driver.configure(
        right_arm_config.model,
        right_arm_config.end_effector,
        right_arm_ip, false
    );
    right_driver.set_all_modes(trossen_arm::Mode::position);

    left_driver.configure(
        left_arm_config.model,
        left_arm_config.end_effector,
        left_arm_ip, false
    );
    left_driver.set_all_modes(trossen_arm::Mode::position);

    std::cout << "Moving arms to start position" << std::endl;
    right_driver.set_arm_positions(START_POSE, 2.0, true);
    left_driver.set_arm_positions(START_POSE, 2.0, true);

    // Network setup
    trossen_vr::NetworkManager receiver(net_config);
    receiver.start();

    // Arm state
    ArmTeleopState right_state, left_state;

    // Track previous tracking state for grip engage/disengage detection
    uint8_t prev_right_tracked = 0;
    uint8_t prev_left_tracked = 0;

    // Event-driven teleop setup
    trossen_vr::Teleop teleop;

    // Button B - exit
    teleop.on_button_b([&]() {
        std::cout << "Exit requested via B button" << std::endl;
        running = false;
    });

    // Right trigger - right gripper
    teleop.on_right_trigger([&](double val) {
        right_driver.set_gripper_position(val * gripper_max_m, 0.0, false);
    });

    // Left trigger - left gripper
    teleop.on_left_trigger([&](double val) {
        left_driver.set_gripper_position(val * gripper_max_m, 0.0, false);
    });

    // Right controller pose - auto engage/disengage based on tracking
    teleop.on_right_pose([&](const trossen_vr::Pose6D& pose) {
        right_state.last_vr_pose = pose;
        right_state.pose_valid = true;

        if (!prev_right_tracked) {
            // Auto-engage arm
            engage_arm(right_state, right_driver);
            std::cout << "Right arm ENGAGED" << std::endl;
        }
        prev_right_tracked = 1;
    });

    // Left controller pose - auto engage/disengage based on tracking
    teleop.on_left_pose([&](const trossen_vr::Pose6D& pose) {
        left_state.last_vr_pose = pose;
        left_state.pose_valid = true;

        if (!prev_left_tracked) {
            // Auto-engage arm
            engage_arm(left_state, left_driver);
            std::cout << "Left arm ENGAGED" << std::endl;
        }
        prev_left_tracked = 1;
    });

    // Main loop
    const double send_period = 1.0 / send_rate_hz;
    auto last_send_time = std::chrono::steady_clock::now();
    trossen_vr::ConnectionStatus last_status = trossen_vr::ConnectionStatus::Disconnected;

    std::cout << "Waiting for VR data... Hand/Grip trigger to engage, release to pause. Press B to exit" << std::endl;

    while (running) {
        // Monitor connection status
        auto current_status = receiver.get_connection_status();
        if (current_status != last_status) {
            switch (current_status) {
                case trossen_vr::ConnectionStatus::Connecting:
                    std::cout << "Connecting..." << std::endl;
                    break;
                case trossen_vr::ConnectionStatus::Connected:
                    std::cout << "Connection established (";
                    std::cout << receiver.get_message_frequency() << " Hz)" << std::endl;
                    break;
                case trossen_vr::ConnectionStatus::Degraded:
                    std::cout << "Connection degraded (low frequency: ";
                    std::cout << receiver.get_message_frequency() << " Hz)" << std::endl;
                    break;
                case trossen_vr::ConnectionStatus::Disconnected:
                    std::cout << "Connection lost (timeout)" << std::endl;
                    break;
            }
            last_status = current_status;
        }

        auto frame = receiver.latest_frame();
        if (!frame) {
            std::this_thread::sleep_for(std::chrono::milliseconds(1));
            continue;
        }

        // Dispatch to event handlers
        teleop.dispatch(*frame);

        // Check for tracking release - disengage
        bool right_tracked = frame->right_controller.is_tracked != 0;
        bool left_tracked = frame->left_controller.is_tracked != 0;

        if (!right_tracked && prev_right_tracked) {
            disengage_arm(right_state);
            std::cout << "Right arm PAUSED" << std::endl;
            prev_right_tracked = 0;
        }
        if (!left_tracked && prev_left_tracked) {
            disengage_arm(left_state);
            std::cout << "Left arm PAUSED" << std::endl;
            prev_left_tracked = 0;
        }

        // Rate-limited command sending
        auto now = std::chrono::steady_clock::now();
        std::chrono::duration<double> elapsed = now - last_send_time;
        if (elapsed.count() >= send_period) {
            send_arm_command(right_state, right_driver, cmd_goal_time);
            send_arm_command(left_state, left_driver, cmd_goal_time);
            last_send_time = now;
        }
    }

    // Shutdown
    receiver.stop();
    std::cout << "\nShutting down..." << std::endl;
    std::cout << "Moving arms to idle position" << std::endl;
    right_driver.set_arm_positions(IDLE_POSE, 2.0, true);
    left_driver.set_arm_positions(IDLE_POSE, 2.0, true);

    return 0;
}
